File: leetcode/119.pascal's triangle II/answer.js
/**
 * @param {number} rowIndex
 * @return {number[]}
 */

var getRow = function (rowIndex) {
  let result = rowIndex > 10 ? new Int32Array(rowIndex + 1) : Buffer.allocUnsafe(rowIndex + 1);
  result[0] = 1;
  if (rowIndex > 0) {
    const vector_index_max = rowIndex;
    const target_row = rowIndex;
    let current_row_to_fill = 1
    let formal_length = 1;
    function fill_row_element() {
      let length =  0|(current_row_to_fill / 2) + 1;
      let write_to_right = current_row_to_fill & 1;
      if (write_to_right) {
        for (let i = 0; i < length; i++) {
          let left = 0;
          if (i > 0) {
            left = result[i - 1];
          }

          let right = left;
          if (i < formal_length) {
            right = result[i];
          }
          result[vector_index_max - i] = left + right;
        }
      }
      else {
        for (let i = 0; i < length; i++) {
          let left = 0;
          if (i > 0) {
            left = result[vector_index_max - i + 1];
          }

          let right = left;
          if (i < formal_length) {
            right = result[vector_index_max - i];
          }
          result[i] = left + right;
        }
      }

      if (current_row_to_fill == target_row) {
        for (let i = 0; i < length; i++) {
          if (write_to_right) {
            result[i] = result[vector_index_max - i];
          }
          else {
            result[vector_index_max - i] = result[i];
          }
        }
        return;
      }
      else {
        formal_length = length;
        current_row_to_fill++;
        return fill_row_element();
      }
    }
    fill_row_element();
  }
  return [...result];
};

File: leetcode/119.pascal's triangle II/answer.cpp
class Solution {
    union {
        int vector_index_max;
        int target_row;
    }u_;
    int formal_length_;
public:
    void fill_row_element(vector<int>&row, int row_to_fill) {
        auto length = row_to_fill / 2 + 1;
        auto write_to_right = row_to_fill & 1;
        if (write_to_right) {
            for (auto i = 0; i < length; i++) {
                auto left = 0;
                if (i > 0) {
                    left = row[i - 1];
                }
                
                auto right = left;
                if (i < formal_length_) {
                    right = row[i];
                }
                row[u_.vector_index_max - i] = left + right;
            }
        }
        else {
            for (auto i = 0; i < length; i++) {
                auto left = 0;
                if (i > 0) {
                    left = row[u_.vector_index_max - i + 1];
                }
                
                auto right = left;
                if (i < formal_length_) {
                    right = row[u_.vector_index_max - i];
                }
                row[i] = left + right;
            }
        }
        
        if (row_to_fill == u_.target_row) {
            for (auto i = 0; i < length; i++) {
                if (write_to_right) {
                    row[i] = row[u_.vector_index_max - i];
                }
                else {
                    row[u_.vector_index_max - i] = row[i];
                }
            }
            return;
        }
        else {
            formal_length_ = length;
            return fill_row_element(row, row_to_fill + 1);
        }
    }
    
    vector<int> getRow(int rowIndex) {
        vector<int> result(rowIndex + 1);
        result[0] = 1;
        
        if (rowIndex > 0) {
            u_.vector_index_max = rowIndex;
            // u_.target_row = rowIndex;
            formal_length_ = 1;
            fill_row_element(result, 1);
        }
        return result;
    }
};
